### PR TITLE
mds: fix cond that prevents sending MLock messages to starting MDS

### PR DESCRIPTION
**Pull request by batrick**
Fixes: http://tracker.ceph.com/issues/23812

Signed-off-by: Patrick Donnelly <pdonnell@redhat.com>

### CODE REVIEW

**Comment by batrick** [comment]
This is not quite right. That actual fix is to change `MDSMap::is_degraded` to return true if a rank is starting. I'll fix that shortly. (Time for dinner!)

File: src/mds/Locker.cc [hunk 1]
```diff
@@ -890,7 +890,7 @@ void Locker::eval_gather(SimpleLock *lock, bool first, bool *pneed_issue, list<M
 	return;
       }
 
-      if (!mds->is_cluster_degraded() ||
+      if (!mds->is_cluster_degraded() &&
 	  mds->mdsmap->get_state(auth) >= MDSMap::STATE_REJOIN) {
```
Comment (ukernel) [inline src/mds/Locker.cc]: it's not wrong. !mds->is_cluster_degraded() check is an optimization, it avoids checking target mds' state for each message

**Comment by batrick** [comment]
@ukernel PTAL. It looks to fix the problem in my testing. Please also look at https://github.com/ceph/ceph/pull/18278#issuecomment-383267965

**Comment by ukernel** [comment]
The root cause is that  mds discovers root inode when it's still in starting state. It should handle lock message after it get replica of root inode.  I think a better to discover root inode when mds is active. (just like the code for creating new mds rank).  The subtle part is that, when mds is in starting state, it needs to load dirfrag of its mdsdir, then creates a new log segments

**Comment by batrick** [comment]
@ukernel, that may be the cause but does it need fixed if we do this patch? It seems to me that having one rank in up:starting is actually a degraded state since the rank is unavailable which could prevent lock processing

**Comment by batrick** [comment]
Actually what I said makes no sense since a `up:starting` MDS wouldn't be holding any locks or authoritative for any subtrees.

Still, I feel "degraded" might not be the right thing here though and perhaps we just need a different term to represent that the MDS cluster has ranks which are creating/starting/recovering.

**Comment by ukernel** [comment]
I think  degraded is OK for cluster with failed or recovering mds.  If we make mds not discover inode when it's in creating/starting state, I think we don't need to create new state for cluster with creating/starting mds.

**Comment by batrick** [comment]
Sure!

**Pull request closed without merging**